### DISCUSSION TITLE
Strange behaviour in data range by local view

### DISCUSSION THREAD

**Discussion by Jimmy-KL** [Kitware/trame#427] [Q&A] [answered]
I found very strange behavior in the data range by the local view and did a simple example to explain it. I think the problem is that I don't know some logic in the local view.
Here is the sample code:
```
from trame.app import get_server
from trame.ui.vuetify3 import SinglePageLayout
from trame.widgets import vuetify3

import pyvista as pv
from pyvista.trame.ui import plotter_ui

# -----------------------------------------------------------------------------
# Trame initialization
# -----------------------------------------------------------------------------
pv.global_theme.allow_empty_mesh = True
pv.OFF_SCREEN = True

server = get_server()
state, ctrl = server.state, server.controller
# state.show = True

# -----------------------------------------------------------------------------
# Pipelines
# -----------------------------------------------------------------------------

mesh = pv.SolidSphere(outer_radius=1)
mesh["data"] = mesh.points[:,0]

pl = pv.Plotter()
mesh_actor = pl.add_mesh(mesh, clim=[mesh["data"].min(), mesh["data"].max()])
slice_mesh = mesh.slice(normal="x")
slice_actor = pl.add_mesh(slice_mesh, clim=[mesh["data"].min(), mesh["data"].max()])

# -----------------------------------------------------------------------------
# Call backs
# -----------------------------------------------------------------------------

@state.change("x_position")
def update_slice(x_position, **kwargs):
    slice_mesh = mesh.slice(normal="x", origin=[x_position, 0, 0])
    slice_actor.mapper.dataset = slice_mesh
    ctrl.view_update()

@state.change("opacity")
def update_mesh_opacity(opacity, **kwargs):
    mesh_actor.prop.opacity = opacity
    ctrl.view_update()

@state.change("show")
def update_mesh_visibility(show, **kwargs):
    mesh_actor.visibility = show
    ctrl.view_update()


# -----------------------------------------------------------------------------
# GUI
# -----------------------------------------------------------------------------

with SinglePageLayout(server) as layout:

    with layout.toolbar:
        vuetify3.VSpacer()
        vuetify3.VSlider(
            v_model=("x_position", 0),
            label="XSlice",
            min=-0.5,
            max=0.5,
            step=0.05,
            dense=True,
            hide_details=True,
        )
        vuetify3.VSlider(
            v_model=("opacity", 1),
            label="Opacity",
            min=0,
            max=1,
            step=0.1,
            dense=True,
            hide_details=True,
        )
        vuetify3.VCheckbox(
            v_model=("show", True),
            label = "Visibility",
            hide_details=True,
            # update_modelValue=()
        )

    with layout.content:
        with vuetify3.VContainer(
            fluid=True,
            classes="pa-0 fill-height",
            style="position: relative;"
        ):
            view = plotter_ui(pl)
            ctrl.view_update = view.update

server.start()
```
Everything is OK in the above example until I use an empty slice mesh when initializing the application by setting the initial x slice position to -5 (the min x coordinate is -1):
```
  v_model=("x_position", 0),
  label="XSlice",
```
By changing the code as above, we will get an empty slice mesh at the beginning, and then strange things happen.
By remote view, everything is alright. As the slice position is at -5 (empty mesh), so it's invisible:
![image](https://github.com/Kitware/trame/assets/114913030/2b2eade9-b2c5-4be6-8c2f-7c2cfe2ec8fc)
![image](https://github.com/Kitware/trame/assets/114913030/0a878c14-6535-4dc1-8081-87e2a22a49f7)

Then let's change to the local view. In the local view, the scalar bar changes to a very strange range - **(0, 250)**. I think this is caused by the empty slice mesh because if there is slice mesh at the initial stage, everything is OK.
![image](https://github.com/Kitware/trame/assets/114913030/96160a83-f648-4eb9-b426-ecbab572195f)
Then, let's move the slice bar to get a non-empty slice mesh. The strange thing happened again, the range became **(0, 1.0)** which should be (-1, 1). If we change to remote view, everything is OK and we can see the color of the slice mesh changed too.
![image](https://github.com/Kitware/trame/assets/114913030/508cf937-b413-4ece-b264-3bef83241667)
We can also see a little difference in the color of the mesh between the remote and local view - the remote view is a little brighter than the local view. The difference is either significant as shown in the scalar bar or exactly the same which is very strange to me.
Local view:
![image](https://github.com/Kitware/trame/assets/114913030/014b625d-d01d-4a7a-a8d9-a854323259a6)
Remote view:
![image](https://github.com/Kitware/trame/assets/114913030/948514bd-cdeb-4df1-bc40-dda191f4ffef)

**Comment by jourdain** [comment, marked as answer]
Sorry could not find the time to really investigate.

Could you try to update `trame-vtk`, add [that line](https://github.com/Kitware/trame-vtk/blob/master/examples/validation/ColorByComponentPyVista.py#L11) in your code and report back?

**Reply by Jimmy-KL** [reply]
Great, the problem has been solved by the new version and `encode_lut(True)`.

**Reply by jourdain** [reply]
Cool so the current default improve network bandwidth but don't always do a good job getting the LUT correctly for vtk.js.
At least now we have a workaround. Thanks for validating it...
